tutorials/preprocessing/plot_35_artifact_correction_regression.py

```diff
@@ -0,0 +1,142 @@
+
+events = mne.find_events(raw, stim_channel='UPPT001')
+sound_data = raw[raw.ch_names.index('UADC001-4408')][0][0]
+onsets = np.where(np.abs(sound_data) > 2. * np.std(sound_data))[0]
+min_diff = int(0.5 * raw.info['sfreq'])
+diffs = np.concatenate([[min_diff + 1], np.diff(onsets)])
+onsets = onsets[diffs > min_diff]
+assert len(onsets) == len(events)
+events[:, 0] = onsets
+epochs = mne.Epochs(raw, events, event_id=1, decim=decim, preload=True)
+
+###############################################################################
+# Now let's regress our EOG signal and plot the original and processed data.
+# We do this by estimating the regression coefficients on data without the
+# evoked response, then use those coefficients to remove the EOG signal:
+
+plot_picks = ['meg', 'eog', 'ecg']
+evo_kwargs = dict(picks=plot_picks, spatial_colors=True,
+                  verbose='error')  # ignore warnings about spatial colors
+fig = epochs.average(picks=plot_picks).plot(**evo_kwargs)
+fig.suptitle('Auditory epochs')
+fig.tight_layout()
+
+epochs_no_ave = epochs.copy().subtract_evoked()
+_, betas = mne.preprocessing.regress_artifact(epochs_no_ave)
+epochs_clean, _ = mne.preprocessing.regress_artifact(epochs, betas=betas)
+fig = epochs_clean.average(picks=plot_picks).plot(**evo_kwargs)
+fig.suptitle('Auditory epochs, EOG regressed')
+fig.tight_layout()
+del epochs, epochs_clean
+
+###############################################################################
+# The effect is subtle in these evoked data. You can see that a bump toward
+# the end of the window has had its amplitude decreased.
+#
+# The effect is even clearer if we create epochs around our (autodetected) EOG
+# events and plot the average across those epochs:
+
+eog_epochs = mne.preprocessing.create_eog_epochs(raw, decim=decim)
+eog_epochs.apply_baseline((None, None))
+order = np.concatenate([  # plotting order: EOG+ECG first, then MEG
+    mne.pick_types(raw.info, meg=False, eog=True, ecg=True),
+    mne.pick_types(raw.info, meg=True)])
+raw_kwargs = dict(order=order, duration=25, n_channels=40)
+raw.plot(events=eog_epochs.events, **raw_kwargs)
+fig = eog_epochs.average(picks=plot_picks).plot(**evo_kwargs)
+fig.suptitle('EOG epochs')
+fig.tight_layout()
+
+###############################################################################
+# And then clean those data:
+
+raw_clean, _ = mne.preprocessing.regress_artifact(raw, betas=betas)
+raw_clean.plot(events=eog_epochs.events, **raw_kwargs)
+eog_epochs_clean, _ = mne.preprocessing.regress_artifact(eog_epochs,
+                                                         betas=betas)
+fig = eog_epochs_clean.average(picks=plot_picks).plot(**evo_kwargs)
+fig.suptitle('EOG epochs, EOG regressed')
+fig.tight_layout()
+del eog_epochs, eog_epochs_clean, raw_clean  # save memory
+
+###############################################################################
+# ECG
+# ~~~
+# We can follow the same process with ECG:
```

Comment:
```suggestion
# We can follow the same process with ECG, although it turns out not to work as
# well. This is likely because the ECG signal is like a rotating dipole, and
# therefore the ECG electrode's time waveform does not reflect the same
# temporal dynamics that manifest at each MEG channel. Other approaches like
# :ref:`ICA <tut-artifact-ica>` or :ref:`SSP <tut-artifact-ssp>` may work
# better with this dataset.
```